osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/DlmsPushNotificationDecoderTest.java

```diff
@@ -0,0 +1,136 @@
+        assertThat(dlmsPushNotification.getAlarms().isEmpty()).isTrue();
+        assertThat(dlmsPushNotification.toByteArray()).isEqualTo(bytes);
+    }
+
+    @Test
+    public void decodeDsmr4AlarmsWithAlarmRegister() throws UnknownDecodingStateException,
+            UnrecognizedMessageDataException {
+
+        // SETUP
+
+        decoder = new DlmsPushNotificationDecoder();
+
+        ChannelBuffer buffer = mock(ChannelBuffer.class);
+        DecodingState state = DecodingState.EQUIPMENT_IDENTIFIER;
+
+        // Create alarm register with 3 alarms: replace battery and 2 mbus alarms
+        final byte[] alarmRegister = new byte[] { 0x00, 0x18, 0x00, 0x02 };
+
+        byte[] bytes = setupDsmr4Buffer(buffer, IDENTIFIER, alarmRegister);
+
+        // CALL
+
+        Object pushNotificationObject = decoder.decode(ctx, channel, buffer, state);
+
+        // VERIFY
+
+        assertThat(pushNotificationObject instanceof DlmsPushNotification).isTrue();
+        DlmsPushNotification dlmsPushNotification = (DlmsPushNotification) pushNotificationObject;
+        assertThat(dlmsPushNotification.getEquipmentIdentifier()).isEqualTo(IDENTIFIER);
+        assertThat(dlmsPushNotification.getTriggerType()).isEqualTo(PUSH_ALARM_TRIGGER);
+        assertThat(dlmsPushNotification.toByteArray()).isEqualTo(bytes);
+
+        Set<AlarmTypeDto> alarms = dlmsPushNotification.getAlarms();
+        assertThat(alarms.size()).isEqualTo(3);
+        assertThat(alarms.contains(REPLACE_BATTERY)).isTrue();
+        assertThat(alarms.contains(COMMUNICATION_ERROR_M_BUS_CHANNEL_4)).isTrue();
+        assertThat(alarms.contains(FRAUD_ATTEMPT_M_BUS_CHANNEL_1)).isTrue();
+    }
+
+    private byte[] setupDsmr4Buffer(ChannelBuffer buffer, String identifier, byte[] data) {
+
+        final byte[] bytes = new byte[EQUIPMENT_IDENTIFIER_LENGTH + COMMA_LENGTH + data.length] ;
+
+        System.arraycopy(identifier.getBytes(), 0, bytes, 0, EQUIPMENT_IDENTIFIER_LENGTH);
+        System.arraycopy(new byte[] { COMMA }, 0, bytes, EQUIPMENT_IDENTIFIER_LENGTH, COMMA_LENGTH);
+        System.arraycopy(data, 0, bytes, EQUIPMENT_IDENTIFIER_LENGTH + COMMA_LENGTH, data.length);
+
+        doAnswer(invocation-> {
+            byte[] outputValue = (byte[])invocation.getArguments()[0];
+            System.arraycopy(bytes, 0, outputValue, 0, EQUIPMENT_IDENTIFIER_LENGTH + 1);
+            return null;
+        }).when(buffer).readBytes(any(byte[].class), eq(0), eq(EQUIPMENT_IDENTIFIER_LENGTH + 1));
+
+        doAnswer(invocation-> {
+            byte[] outputValue = (byte[])invocation.getArguments()[0];
+            System.arraycopy(bytes, EQUIPMENT_IDENTIFIER_LENGTH + 1, outputValue, 0, data.length);
+            return null;
+        }).when(buffer).readBytes(any(byte[].class), eq(0), eq(data.length));
+
+        when(buffer.writerIndex()).thenReturn(bytes.length);
+        when(buffer.readerIndex()).thenReturn(EQUIPMENT_IDENTIFIER_LENGTH + 1);
```

Comment:
I see some when statements, but no verify statements. Each when statement should have a matching verify to make sure the method is called in the test.

Reply:
Ok, added verify for each when

Reply:
it's not true that you must have a verify for each when statement. When statements are intended to stub behavior needed for the test case to execute as desired. Verify statements are needed to check that any expected behaviors on mocks have occurred. You can have verify statements without when statements and vice versa. Indeed, it is bad practice to add redundant verify statements to check that the mock has been invoked, if the fact that the correct, non-null value has been returned by the mock is already obvious from the successful completion of the case.

Reply:
this discussion pops up every project. We can agree to not verify if that is the consensus within the project. Last project we decided to verify always.

Reply:
Before we reach consensus I always like to hear the reasons. What value does it add to have this rule?